crates/miden-objects/src/account/component/metadata/storage/value_name.rs

```diff
@@ -0,0 +1,140 @@
+use alloc::string::{String, ToString};
+use core::fmt::{self, Display};
+use core::str::FromStr;
+
+use miden_core::utils::{ByteReader, ByteWriter, Deserializable, Serializable};
+use miden_processor::DeserializationError;
+use thiserror::Error;
+
+use crate::account::StorageSlotName;
+
+/// A simple wrapper type around a string key that identifies init-provided values.
+///
+/// A storage value name is a string that identifies values supplied during component
+/// instantiation (via [`InitStorageData`](super::InitStorageData)).
+///
+/// Names can be chained together to form unique keys for nested schema fields (e.g. a composed
+/// word slot with per-felt typed fields).
+#[derive(Clone, Debug, Ord, PartialOrd, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(::serde::Deserialize, ::serde::Serialize))]
+#[cfg_attr(feature = "std", serde(transparent))]
+pub struct StorageValueName {
+    fully_qualified_name: String,
+}
+
+impl StorageValueName {
+    /// Creates a [`StorageValueName`] for the given storage slot.
+    ///
+    /// This is an infallible conversion: a [`StorageSlotName`] is always a valid storage value
+    /// name prefix.
+    pub fn from_slot_name(slot_name: &StorageSlotName) -> Self {
+        StorageValueName {
+            fully_qualified_name: slot_name.as_str().to_string(),
+        }
+    }
+
+    /// Creates a [`StorageValueName`] for a named field within a storage slot.
+    ///
+    /// This is equivalent to `StorageValueName::from_slot_name(slot).with_suffix(field)`.
+    pub fn from_slot_name_and_suffix(
+        slot_name: &StorageSlotName,
+        suffix: &StorageValueName,
+    ) -> Self {
+        Self::from_slot_name(slot_name).with_suffix(suffix)
+    }
+
+    /// Creates a [`StorageValueName`] for a named field within a storage slot.
+    pub fn from_slot_name_and_field(
+        slot_name: &StorageSlotName,
+        field: impl Into<String>,
+    ) -> Result<Self, StorageValueNameError> {
+        let field: String = field.into();
+        let suffix: StorageValueName = field.parse()?;
+        Ok(Self::from_slot_name_and_suffix(slot_name, &suffix))
+    }
+
+    /// Creates an empty [`StorageValueName`].
+    pub(crate) fn empty() -> Self {
+        StorageValueName { fully_qualified_name: String::default() }
+    }
+
+    /// Adds a suffix to the storage value name, separated by a period.
+    #[must_use]
+    pub fn with_suffix(self, suffix: &StorageValueName) -> StorageValueName {
```

Comment:
I wonder if we should change the signature here to take `impl Into<String>` as the second parameter. I guess we are saving a bit on validation code if we pass `StorageValueName` - but it also feels a bit odd that `StorageValueName` can be its own suffix.

Reply:
Same as [here](https://github.com/0xMiden/miden-base/pull/2193#discussion_r2632853546)